### PR TITLE
Complete from CT, pictures, birthday ...

### PR DESCRIPTION
**Pull request by Rany615**
big update: mail and hierarchy added to DU, added pictures to kartoffel, added birthday and sex to soldiers from aka and kartoffel,
and complete from cts if complete from aka did not exist.

still needs: match the unit names from CT to aka format

### CODE REVIEW

File: mocks/mockGen.js [hunk 1]
```diff
@@ -183,8 +190,51 @@ for (let i = 0; i < miriAmount; i++) {
     miriUsers.push(miriUser);
 }
 
+// Generating pictures
+for (let i = 0; i < picturesAmount; i++) {
+    let picture = {}
+    picture.profile = {};
```
Comment (NoamYes) [inline mocks/mockGen.js:196]: is it in profile or profiles and avatars under different path

File: mocks/mockGen.js [hunk 2]
```diff
@@ -183,8 +190,51 @@ for (let i = 0; i < miriAmount; i++) {
...
+    picture.profile.personalNumber = mis[i];
+    picture.profile.path = utils.generateNumberBody();
+    picture.profile.takenAt = utils.randomElement(
```
Comment (NoamYes) [inline mocks/mockGen.js:199]: add format field

File: mocks/mockGen.js [hunk 3]
```diff
@@ -183,8 +190,51 @@ for (let i = 0; i < miriAmount; i++) {
...
+        [
+          faker.date
+            .between(faker.date.future(10), faker.date.past(10))
```
Comment (NoamYes) [inline mocks/mockGen.js:202]: prefer real dates

File: src/config/fieldNames.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ module.exports = {
         mdn: "mdn_name",
         mm: "mm_name",
         city: "city_name",
-        oa: "oa",
+        pictures: "pictures",
```
Comment (NoamYes) [inline src/config/fieldNames.js:13]: remove because from aka

File: src/config/fieldNames.js [hunk 5]
```diff
@@ -29,6 +29,8 @@ module.exports = {
         unitName: "hr", 
         telephoneType: "telephoneType",
         uniqeFieldForDeepDiff: "mi",
+        birthday: "moledet",
```
Comment (NoamYes) [inline src/config/fieldNames.js:32]: change to birth

File: src/config/paths.js [hunk 6]
```diff
@@ -15,6 +15,7 @@ module.exports = pathsHandler = (param, param2) => {
         [`${fn.dataSources.mdn}_API`]: 'http://localhost:3001/getNv/sql/mdn',
         [`${fn.dataSources.city}_API`]: 'http://localhost:3001/getCity',
         immediateServer_API: '/immediateRun',
+        PICTURES_API: "",
```
Comment (NoamYes) [inline src/config/paths.js:18]: same as getAkaTelephone -> getAkaImgMetaData

File: src/runningMethods/immediate.js [hunk 7]
```diff
@@ -1,6 +1,7 @@
 const fn = require('../config/fieldNames');
 const diffsHandler = require('../util/diffsHandler');
 const preRun = require('../util/preRun');
+const cleanImmediateDus = require('../util/cleanDu/cleanImmediateDus');
```
Comment (NoamYes) [inline src/runningMethods/immediate.js:4]: clean du

File: src/runningMethods/immediate.js [hunk 8]
```diff
@@ -15,10 +16,18 @@ module.exports = async (dataSource, identifiersArray, runUID) => {
 
         let akaRecords = dataObj[fn.dataSources.aka] ? dataObj[fn.dataSources.aka].data : [];
         let foundRecords = dataObj[dataSource] ? dataObj[dataSource].data : [];
-
         const Auth = new AuthClass(sendLog);
 
         await diffsHandler({ added: foundRecords }, dataSource, akaRecords, fn.runnigTypes.immediateRun, sendLog, Auth);
+        if(dataSource != fn.dataSources.aka) {
+            await cleanImmediateDus(
```
Comment (NoamYes) [inline src/runningMethods/immediate.js:23]: remove cleandu

File: src/util/completeFromAka.js [hunk 9]
```diff
@@ -3,6 +3,7 @@ const validators = require('../config/validators');
 const { logLevel } = require('./logger');
 const logDetails = require('../util/logDetails');
 const mergeArrays = require('./generalUtils/mergeArrays');
+const moment = require('moment');
```
Comment (NoamYes) [inline src/util/completeFromAka.js:6]: check if needed

File: src/util/completeFromCT.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+const fn = require('../config/fieldNames');
+const { logLevel } = require('./logger');
+const logDetails = require('../util/logDetails');
+const moment = require('moment');
```
Comment (NoamYes) [inline src/util/completeFromCT.js:4]: check if needed

File: src/util/completeFromCT.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * This module add fields from city dataSource to given object
+ *
+ * @param {*} obj Object of person
+ * @param {*} CTRecord city dataSource
+ * @returns Object of person with the data from city
+ */
+module.exports =  async (obj, CTRecord, sendLog) => {
+    let identifier = obj.personalNumber || obj.identityCard;
```
Comment (NoamYes) [inline src/util/completeFromCT.js:14]: the identifier may be unnecessary

File: src/util/diffsHandler.js [hunk 12]
```diff
@@ -12,14 +12,15 @@ require('dotenv').config();
...
     const updatedData = updated.map(
         deepDiffObj => new DataModel(deepDiffObj[1], dataSource, fn.flowTypes.update, runnigType, Auth, sendLog, deepDiffObj)
     );
-
+    
+    
     return PromiseAllWithFails([
-        add({ addedData, dataSource }, aka_all_data),
+        add({ addedData, dataSource }, aka_all_data , ct_all_data , pictures_all_data),
         update({ updatedData, dataSource }, aka_all_data),
```
Comment (NoamYes) [inline src/util/diffsHandler.js:24]: check if needed in update ct also

File: src/util/completeFromAka.js [hunk 13]
```diff
@@ -110,7 +122,7 @@ function phonesValueHandler(person, phone, phoneType) {
  * @param {*} dataSource The dataSource of the person object
  * @returns Object of person with the data from aka
  */
-module.exports = (obj, akaData, dataSource, sendLog) => {
+module.exports =  async (obj, akaData, dataSource, sendLog) => {
```
Comment (NoamYes) [inline src/util/completeFromAka.js:125]: remove async

File: src/util/completeFromCT.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+const moment = require('moment');
+
+/**
+ * This module add fields from city dataSource to given object
+ *
+ * @param {*} obj Object of person
+ * @param {*} CTRecord city dataSource
+ * @returns Object of person with the data from city
+ */
+module.exports =  async (obj, CTRecord, sendLog) => {
```
Comment (NoamYes) [inline src/util/completeFromCT.js:13]: async probably not needed

File: src/util/diffsHandlerUtils/addedDataHandler.js [hunk 15]
```diff
@@ -37,7 +40,23 @@ module.exports = async ({ addedData, dataSource }, aka_all_data) => {
             DataModel.person_ready_for_kartoffel.entityType === fn.entityTypeValue.s ||
             DataModel.person_ready_for_kartoffel.entityType === fn.entityTypeValue.c
         ) {
-            DataModel.completeFromAka(aka_all_data);
+            let identifier = DataModel.person_ready_for_kartoffel.personalNumber || DataModel.person_ready_for_kartoffel.identityCard;
+            let akaRecord = aka_all_data.find(person => ((person[fn.aka.personalNumber] == identifier) || (person[fn.aka.identityCard] == identifier)));
+            if(akaRecord){
+                await PromiseAllWithFails(DataModel.completeFromAka(aka_all_data));
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/addedDataHandler.js:46]: if not async no need for promise

File: src/util/diffsHandlerUtils/addedDataHandler.js [hunk 16]
```diff
@@ -37,7 +40,23 @@ module.exports = async ({ addedData, dataSource }, aka_all_data) => {
...
+            }
+            else{
+                let CTRecord = ct_all_data.find(person => ((person[fn.city_name.personalNumber] == identifier) || (person[fn.city_name.identityCard] == identifier)));
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/addedDataHandler.js:49]: pay attention to whether can be replaced by const

File: src/util/fieldsUtils/domainUserHandler.js [hunk 17]
```diff
@@ -16,10 +16,11 @@ const assembleDomainUser = require('./assembleDomainUser');
 module.exports = async (DataModel) => {
     let user_object = {
         dataSource: DataModel.dataSource,
+        mail: DataModel.person.mail,
```
Comment (NoamYes) [inline src/util/fieldsUtils/domainUserHandler.js:19]: add a comment to all hierarchy per domain user

File: src/util/matchToKartoffel.js [hunk 18]
```diff
@@ -191,6 +191,16 @@ const match_es = (obj, dataSource) => {
                 obj.job = location ? `${job} - ${location}` : job;
                 (rawKey === "job") ? null : delete obj[rawKey];
                 break;
+            //Birthday
+            case fn[dataSource].birthday:
```
Comment (NoamYes) [inline src/util/matchToKartoffel.js:195]: move to match_aka

**Review by NoamYes — COMMENTED** [review]
- Remove karting.zip
- Remove cleanDUs files and stuff - cleanImmediateDus.js

File: src/util/completeFromAka.js [hunk 19]
```diff
@@ -83,6 +92,8 @@ const complete_city = (obj, akaRecord) => {
     obj.serviceType = akaRecord[fn.aka.serviceType];
     obj.personalNumber = akaRecord[fn.aka.personalNumber];
     obj.entityType = fn.entityTypeValue.s;
+    obj.birthday = akaRecord[fn.aka.birthday];
+    obj.sex = akaRecord[fn.aka.sex];
```
Comment (Eli-BB8) [inline src/util/completeFromAka.js:96]: check if works

File: src/util/completeFromAka.js [hunk 20]
```diff
@@ -141,9 +152,6 @@ module.exports = (obj, akaData, dataSource, sendLog) => {
                     sendLog(logLevel.error, logDetails.error.ERR_DATA_SOURCE);
             }
         }
-        else {
```
Comment (Eli-BB8) [inline src/util/completeFromAka.js:144]: send log in wrapping function of data compilation

File: src/util/completeFromCT.js [hunk 21]
```diff
@@ -0,0 +1,18 @@
+const fn = require('../config/fieldNames');
+
+/**
+ * This module add fields from city dataSource to given object
+ *
+ * @param {*} obj Object of person
+ * @param {*} CTRecord city dataSource
+ * @returns Object of person with the data from city
+ */
+module.exports =  (obj, CTRecord) => {
```
Comment (Eli-BB8) [inline src/util/completeFromCT.js:10]: CT > city

File: src/util/completeFromCT.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    
+    obj.rank = CTRecord[fn.city.rank];
```
Comment (Eli-BB8) [inline src/util/completeFromCT.js:12]: add identityCard currentUnit and every other possible field - like aka

File: src/util/diffsHandler.js [hunk 23]
```diff
@@ -12,14 +12,15 @@ require('dotenv').config();
 * aka_all_data - object that contain all the recent data from aka
 */
 
-module.exports = async ({ added = [], updated = [] }, dataSource, aka_all_data, runnigType, sendLog, Auth) => {
+module.exports = async ({ added = [], updated = [] }, dataSource, aka_all_data, ct_all_data, pictures_all_data, runnigType, sendLog, Auth) => {
```
Comment (Eli-BB8) [inline src/util/diffsHandler.js:15]: 1. merge aka_all_data and ct_all_data into extraData
2. pictures_all_data goes into aka_all_data, and taking care of in completefromAka

File: src/util/diffsHandler.js [hunk 24]
```diff
@@ -12,14 +12,15 @@ require('dotenv').config();
...
         deepDiffObj => new DataModel(deepDiffObj[1], dataSource, fn.flowTypes.update, runnigType, Auth, sendLog, deepDiffObj)
     );
-
+    
+    
     return PromiseAllWithFails([
-        add({ addedData, dataSource }, aka_all_data),
-        update({ updatedData, dataSource }, aka_all_data),
+        add({ addedData, dataSource }, aka_all_data , ct_all_data , pictures_all_data),
+        update({ updatedData, dataSource }, aka_all_data , ct_all_data),
```
Comment (Eli-BB8) [inline src/util/diffsHandler.js:24]: comment

File: src/util/completeFromAka.js [hunk 25]
```diff
@@ -14,10 +14,13 @@ const complete_es = (obj, akaRecord) => {
     obj.rank = akaRecord[fn.aka.rank];
```
Comment (Eli-BB8) [inline src/util/completeFromAka.js:14]: pictures will be here

File: src/util/diffsHandlerUtils/addedDataHandler.js [hunk 26]
```diff
@@ -37,7 +40,23 @@ module.exports = async ({ addedData, dataSource }, aka_all_data) => {
             DataModel.person_ready_for_kartoffel.entityType === fn.entityTypeValue.s ||
             DataModel.person_ready_for_kartoffel.entityType === fn.entityTypeValue.c
         ) {
-            DataModel.completeFromAka(aka_all_data);
+            const identifier = DataModel.person_ready_for_kartoffel.personalNumber || DataModel.person_ready_for_kartoffel.identityCard;
+            const akaRecord = aka_all_data.find(person => ((person[fn.aka.personalNumber] == identifier) || (person[fn.aka.identityCard] == identifier)));
```
Comment (Eli-BB8) [inline src/util/diffsHandlerUtils/addedDataHandler.js:44]: move find into DataModel constructor (aka,city)

File: src/util/diffsHandlerUtils/addedDataHandler.js [hunk 27]
```diff
@@ -37,7 +40,23 @@ module.exports = async ({ addedData, dataSource }, aka_all_data) => {
...
+                DataModel.completeFromAka(aka_all_data);
+            }
+            else{
+                const CTRecord = ct_all_data.find(person => ((person[fn.city_name.personalNumber] == identifier) || (person[fn.city_name.identityCard] == identifier)));
+                if(CTRecord){
+                    DataModel.completeFromCT(CTRecord);
+                }
+            }
+            
+            const picturesRecord = pictures_all_data.find(person => ((person[personalNumber] == identifier)));
```
Comment (Eli-BB8) [inline src/util/diffsHandlerUtils/addedDataHandler.js:55]: goes into completefromAka

File: src/util/diffsHandlerUtils/updatedDataHandler.js [hunk 28]
```diff
@@ -16,7 +16,7 @@ require('dotenv').config();
  * @param {string} dataSource - represents the data source
  * @param {Object} aka_all_data - object that contain all the recent data from aka
  */
```
Comment (Eli-BB8) [inline src/util/diffsHandlerUtils/updatedDataHandler.js:18]: extraData..........................

File: src/util/completeFromAka.js (hunk 20 above)
Comment (Rany615) [reply src/util/completeFromAka.js:144]: by log you mean the warning that it wasnt completed by either aka or city?, if so then done

File: src/util/completeFromCT.js (hunk 22 above)
Comment (Rany615) [reply src/util/completeFromCT.js:12]: aka: {
        serviceType: "nstype",
        firstName: "firstName",
        lastName: "lastName",
        identityCard: "tz",
        personalNumber: "mi",
        rank: "rnk",
        phone: "telephone",
        areaCode: "ktelephone",
        mobilePhone: "telephone",
        areaCodeMobile: "ktelephone",
        dischargeDay: "rld",
        clearance: "clearance",
        unitName: "hr", 
        telephoneType: "telephoneType",
        uniqeFieldForDeepDiff: "mi",
        birthday: "birthday",
        sex: "sex",
        picture: "picture",
    },

 city_name: {
        firstName: 'firstName',
        lastName: 'lastName',
        identityCard: 'tz',
        rank: 'rank',
        dischargeDay: 'rld',
        clearance: 'clearance',
        currentUnit: 'department',
        serviceType: 'stype',
        mobilePhone: 'telephone',
        personalNumber: 'personalNumber',
        mail: 'mail',
        job: 'job',
        profession: 'profession',
        hierarchy: 'hr',
        domainUsers: 'domUser',
        company: 'company',
        entityTypePrefix: {
            s: ['e', 'l', 'i'],
            c: ['g', 'j'],
            gu: ['m'],
        },

which ones to be exact? first name and last name too for example?

File: src/runningMethods/recovery.js [hunk 29]
```diff
@@ -17,15 +17,19 @@ module.exports = async () => {
         fn.dataSources.lmn, 
         fn.dataSources.mdn, 
         fn.dataSources.mm, 
-        fn.dataSources.city
-        
+        fn.dataSources.city,
+        fn.dataSources.pictures
     ]);
```
[the remark this answers is not in the document]
Comment (Rany615) [reply src/runningMethods/recovery.js:22]: in akaDataManipulate, check how the code works, and if its correct for pictures

File: mocks/mockGen.js [hunk 30]
```diff
@@ -183,8 +189,55 @@ for (let i = 0; i < miriAmount; i++) {
     miriUsers.push(miriUser);
 }
 
+// Generating pictures
+for (let i = 0; i < picturesAmount; i++) {
+    let picture = {}
+    picture.profile = {};
+    picture.profile.personalNumber = mis[i];
```
Comment (NoamYes) [inline mocks/mockGen.js:196]: should remove profile field

File: src/util/completeFromCity.js [hunk 31]
```diff
@@ -0,0 +1,18 @@
...
+ * This module add fields from city dataSource to given object
+ *
+ * @param {*} obj Object of person
+ * @param {*} CityRecord city dataSource
+ * @returns Object of person with the data from city
+ */
+module.exports =  (obj, CityRecord) => {
+    
+    obj.rank = CityRecord[fn.city.rank];
+    obj.currentUnit = CityRecord[fn.city.unitName];
```
Comment (NoamYes) [inline src/util/completeFromCity.js:13]: comment for now, we dont change unit

File: src/util/matchToKartoffel.js [hunk 32]
```diff
@@ -339,6 +383,21 @@ const match_adNN = (obj, dataSource) => {
...
+                (rawKey === "birthday") ? null : delete obj[rawKey];
+                break;
+            //sex
+            case fn[dataSource].sex:
+                obj.sex = obj[rawKey];
+                (rawKey === "sex") ? null : delete obj[rawKey];
+                break;
+            //picture
+            case fn[dataSource].picture:
+                obj.picture = {"profile" : {"personalNumber" : obj[rawKey].Path,"takenAt" : obj[rawKey].takenAt,"updatedAt" : obj[rawKey].updatedAt,"createdAt" : obj[rawKey].createdAt,"format" : obj[rawKey].format}}
```
Comment (NoamYes) [inline src/util/matchToKartoffel.js:398]: why need in matchToKartoffel? should be removed from all datasources and be in completeFromExtra (?)

**Review by NoamYes — COMMENTED** [review]
Some first notes

**Review by tata406 — APPROVED** [review]

**Review by tata406 — DISMISSED** [review]
Approve

File: src/index.js [hunk 33]
```diff
@@ -3,7 +3,7 @@ const schedule = require("node-schedule");
 const immediateApp = require("./server/immediateServer");
 const recovery = require("./runningMethods/recovery");
 const daily = require("./runningMethods/daily");
-const migrationConsumer = require('./kafka/migration');
+//const migrationConsumer = require('./kafka/migration');
```
Comment (NoamYes) [inline src/index.js:6]: add back

File: src/util/completeFromCity.js [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+ * This module add fields from city dataSource to given object
+ *
+ * @param {*} obj Object of person
+ * @param {*} CityRecord city dataSource
+ * @returns Object of person with the data from city
+ */
+module.exports =  (obj, CityRecord) => {
+    
+    obj.rank = CityRecord[fn.city_name.rank];
+    obj.currentUnit = CityRecord[fn.city_name.unitName];
```
Comment (NoamYes) [inline src/util/completeFromCity.js:13]: comment unit for now

File: src/config/paths.js [hunk 35]
```diff
@@ -7,6 +7,7 @@ module.exports = pathsHandler = (param, param2) => {
         // Data source
         AKA_TELEPHONES_API: "http://localhost:3001/getAkaTelephone",
         AKA_EMPLOYEES_API: "http://localhost:3001/getAkaEmployees",
+        PICTURES_API: "http://localhost:3001/getAkaImgMetaData",
```
Comment (NoamYes) [inline src/config/paths.js:10]: AKA API

File: src/index.js [hunk 36]
```diff
@@ -16,10 +16,10 @@ const scheduleTime =
...
     : new Date().setMilliseconds(new Date().getMilliseconds() + 200);
 
-// schedule.scheduleJob(scheduleTime, async () =>  await daily());
+ //schedule.scheduleJob(scheduleTime, async () =>  await daily());
 schedule.scheduleJob(scheduleRecoveryTime, async () => await recovery());
 
 const port = fn.immediatePort;
 immediateApp.listen(port, () => console.log("immediateRun server run on port:" + port));
 
-migrationConsumer()
```
Comment (NoamYes) [inline src/index.js:25]: remove comments migration

File: src/util/completeFromCity.js (hunk 34 above)
Comment (NoamYes) [inline src/util/completeFromCity.js:13]: comment unit

File: src/util/diffsHandlerUtils/addedDataHandler.js [hunk 37]
```diff
@@ -151,7 +151,7 @@ module.exports = async ({ addedData, dataSource }, aka_all_data) => {
                 if (DataModel.updateDeepDiff && DataModel.updateDeepDiff.length > 0) {
                     updated(
                         { updatedData: [DataModel], dataSource },
-                        aka_all_data
+                        extraData.aka_all_data
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/addedDataHandler.js:154]: to extraData

File: src/util/diffsHandlerUtils/updatedDataHandler.js [hunk 38]
```diff
@@ -88,30 +89,56 @@ module.exports = async ({ updatedData, dataSource }, aka_all_data) => {
...
+            else{
+                const CityRecord = extraData.ct_all_data.find(
+                    person => (
+                        person[fn.city_name.personalNumber] == tryFindPerson.argument ||
+                        person[fn.city_name.identityCard] == tryFindPerson.argument
+                    )
+                );
 
-                    return !include;
+                if(CityRecord &&
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/updatedDataHandler.js:100]: check if need to check primary dataSOurce from city

File: src/util/diffsHandlerUtils/updatedDataHandler.js [hunk 39]
```diff
@@ -88,30 +89,56 @@ module.exports = async ({ updatedData, dataSource }, aka_all_data) => {
...
+                    
+            //         const include = fn.akaRigid.includes(keyForCheck);
+            //         if (include) {
+            //             DataModel.sendLog(
+            //                 logLevel.warn,
+            //                 logDetails.warn.WRN_AKA_FIELD_RIGID,
+            //                 diffsObj.path.toString(),
+            //                 tryFindPerson.argument,
+            //                 DataModel.dataSource
+            //             )
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/updatedDataHandler.js:136]: make sure it is the required state

File: src/util/matchToKartoffel.js [hunk 40]
```diff
@@ -620,9 +619,13 @@ module.exports = async (origin_obj, dataSource, Auth, defaultSendLog, flowType)
             };
             break;
         case fn.dataSources.adNN:
+            //console.log("BBBBB")
```
Comment (NoamYes) [inline src/util/matchToKartoffel.js:622]: remove comments

File: src/util/matchToKartoffel.js [hunk 41]
```diff
@@ -324,9 +323,10 @@ const match_adNN = (obj, dataSource) => {
                 (rawKey === "hierarchy") ? null : delete obj[rawKey];
                 break;
             //personalNumber or identity card
-            case fn[dataSource].sAMAccountName:
+            case fn[dataSource].sAMAccountName:  //changed from sama account, right?
```
Comment (NoamYes) [inline src/util/matchToKartoffel.js:326]: remove comments

**Review by NoamYes — COMMENTED** [review]
add karting.zip to gitignore

File: src/util/diffsHandlerUtils/updatedDataHandler.js [hunk 42]
```diff
@@ -62,7 +62,7 @@ module.exports = async ({ updatedData, dataSource }, aka_all_data) => {
         if (DataModel.dataSource === fn.dataSources.aka) {
```
Comment (NoamYes) [inline src/util/diffsHandlerUtils/updatedDataHandler.js:62]: no need. remove aka flow

**Review by NoamYes — COMMENTED** [review]
add gitignore kartingzip

**Review by NoamYes — APPROVED** [review]
Good luck

**Pull request merged**